src/Database/LSMTree/Internal.hs

```diff
@@ -1278,3 +1304,141 @@ duplicate t@Table{..} = do
+              V.forM ts $ \t -> do
+                withOpenTable t $ \tEnv ->
+                  -- The table contents escape the read access, but we just added references
+                  -- to each run so it is safe.
+                  RW.withReadAccess (tableContent tEnv) (duplicateTableContent reg)
+
+            content <-
+              error "unions: combine contents into merging tree" $ -- TODO
+                contents
+
+            t <-
+              newWith
+                reg
+                sesh
+                seshEnv
+                conf
+                (error "unions: ArenaManager") -- TODO
+                content
+
+            pure (seshState, t)
+  where
+    n = V.length ts
+
+-- | Like 'vmatchBy', but the match function is @(==)@.
+vmatch :: Eq a => V.Vector a -> Either (Int, Int) a
+vmatch = vmatchBy (==)
+
+-- | Check that all values in the vector match. If so, return the matched value.
+-- If there is a mismatch, return the vector indices of the mismatching values.
+--
+-- Assumes the input vector is non-empty.
+vmatchBy :: (a -> a -> Bool) -> V.Vector a -> Either (Int, Int) a
+vmatchBy cmp xs0 =
+    case V.uncons xs0 of
+      Nothing ->
+        error "vmatch: empty vector "
+      Just (x, xs) ->
+        case V.iforM_ xs $ vmatchOne x of
+          Left i   -> Left (0, i)
+          Right () -> Right x
+  where
+    vmatchOne x i y =
+      if (x `cmp` y)
+        then Left i
+        else Right ()
+
+-- | Check that all tables in the session match. If so, return the matched
+-- session. If there is a mismatch, return the vector indices of the mismatching
+-- tables.
+--
+-- Assumes the input vector is non-empty.
+--
+-- TODO: compare LockFileHandle instead of SessionRoot (?). We can write an Eq
+-- instance for LockFileHandle based on pointer equality, just like base does
+-- for Handle.
+vmatchSessions ::
+     (MonadSTM m, MonadThrow m)
+  => V.Vector (Table m h)
+  -> m (Either (Int, Int) (Session m h))
+vmatchSessions ts0
+  | Just (t, ts) <- V.uncons ts0
+  = do
+      root <- getSessionRoot t
+      eith <- go root 1 ts
+      pure $ case eith of
+        Left i   -> Left (0, i)
+        Right () -> Right (tableSession t)
+  | otherwise
+  = error "vmatchSessions: empty vector"
+  where
+    -- Check that the session roots for all tables are the same. There can only
+    -- be one *open/active* session per directory because of cooperative file
+    -- locks, so each unique *open* session has a unique session root.
+    --
+    -- In remote cases, the session can be be closed, after which a new session
+    -- can be opened in the same root directory. In this case, the session root
+    -- is technically not unique. This does not mean that tables in @ts0@ could
+    -- have different sessions with the same root, however. You could prove this
+    -- by contradiction, assuming that there are two tables in @ts0@ that have
+    -- different open sessions with the same session root.
```

Comment:
I'm not totally sure I follow this. We only have to deal with open sessions, and the thing about mutually exclusivity is true for open sessions.

So provided we check the sessions are open (so the file lock is held) then checking it's the same file is surely sufficient?

Reply:
You're right, it's sufficient. This text is a bad attempt at explaining *why* it is sufficient :stuck_out_tongue: 

In general, sessions can have the same session root, as long as the lifetime of the sessions don't overlap. One open session has to be closed before the next one can be opened in the same root. So, in theory, you can have multiple tables in different sessions that have the same root, depending on when you check the root. This is one example:

1. open session 1
2. create table 1 in session 1
3. check the root of table 1's session
4. close session 1
5. open session 2
6. create table 2 in session 2
7. check the root of table 2's session
8. close session 2

This edge case is not possible in our code because of some invariants we have in our code, like how sessions are never re-opened after they are closed, and that the session root never changes, and only one session can have a file lock in a directory. However, it becomes quite convoluted to explain that. So, instead I tell the reader to prove it for themselves. Maybe I should be more brief, and put the text "this check is sufficient, you can prove it for yourself using contradiction" or something similar. Or, maybe I should fully write out an (informal) proof. Or, maybe it's believable enough that checking the session root is sufficient, and I should just leave out the text altogether. Let me know what you think